packages/belos/src/BelosCGSingleRedIter.hpp

```diff
@@ -441,71 +496,145 @@ class CGSingleRedIter : virtual public CGIteration<ScalarType,MV,OP> {
     ////////////////////////////////////////////////////////////////
     // Iterate until the status test tells us to stop.
     //
-    while (1) {
-      
-      // Update the solution vector x := x + alpha * P_
-      //
-      MVT::MvAddMv( one, *cur_soln_vec, alpha, *P_, *cur_soln_vec );
-      lp_->updateSolution();
+    if (foldConvergenceDetectionIntoAllreduce_) {
```

Comment:
Not a huge fan of the massive code duplication here.

Reply:
I can certainly pull the if-statement into the loop. I think it's easier to read that way, since there are a couple of changes between the two paths..